src/libraries/Native/Unix/System.Native/pal_process.c

```diff
@@ -871,3 +861,39 @@ int32_t SystemNative_SchedGetAffinity(int32_t pid, intptr_t* mask)
     return result;
 }
 #endif
+
+// Returns the full path to the executable for the current process resolving symbolic links.
+// The caller is responsible for releasing the buffer. Returns null on error.
+char* SystemNative_GetProcessPath()
+{
+    // Get path to the executable for the current process using
```

Comment:
There are quite a few variants how to implement this, and none of them is obviously the best one.

I ended up taking https://github.com/dotnet/corert/blob/master/src/Native/System.Private.CoreLib.Native/pal_exepath.cpp#L20 as the baseline for this implementation

Reply:
IMVHO, among the bests in terms of completeness, this is the worse. We did more complete implementations of the same exact thing in three other places: https://github.com/dotnet/runtime/blob/cd4cc97e4c099f637061afe2b6c546483ffd3073/src/coreclr/src/hosts/unixcoreruncommon/coreruncommon.cpp#L53-L67 https://github.com/dotnet/runtime/blob/cd4cc97e4c099f637061afe2b6c546483ffd3073/src/installer/corehost/cli/hostmisc/pal.unix.cpp#L807-L890  https://github.com/dotnet/runtime/blob/cd4cc97e4c099f637061afe2b6c546483ffd3073/src/coreclr/src/pal/src/init/pal.cpp#L1300-L1408

Those three places have same parity  in terms of supported base systems. This one is the least complete implementation.

It would be nice to update this one, until we try to converge the four methods into one.

Reply:
Maybe we can put it in `src/shared/exepath.c`, and include that file in projects that need it? Until we have a mini-PAL that you mentioned. If that’s fine, I can work on that. :)

Reply:
I have looked at those other implementations. I see number of places in them that depend on current directory. Depending on current directory maybe acceptable during startup, but it is not acceptable for this implementation that gets called half-way through the process lifetime. It would be a bad experience if changing current directory breaks this API.

Is there a way to create an implementation for all relevant platforms that is not sensitive to current directory changing? If not, maybe this API should be implemented as part of the runtime and the runtime should cache the path during startup.

Reply:
The dependency on cwd (in our code) is present only in the fallback path in case of SunOS. [`getexecname(3C)`](https://illumos.org/man/3c/getexecname). The manpage says:

```
       Normally this is an absolute pathname, as the majority of commands are
       executed by the shells that append the command name to the user's PATH
       components.  If this is not an absolute path, the output of getcwd(3C)
       can be prepended to it to create an absolute path, unless the process or
       one of its ancestors has changed its root directory or current working
       directory since the last successful call to one of the exec family of
       functions.
```

the rare case mentioned is not easily reproduced. I think it requires one of the exec() family of function to dynamically load executable in a way we overwrite the `argv[0]` to NULL or some such. Current implementation is not thread safe.

On macOS's `_NSGetExecutablePath` (which we are using) internally uses cwd and prepends with exec name. It requires us to call `realpath` on the result (which we are not doing). [`dyld(3)`](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dyld.3.html) manual says:

```
     _NSGetExecutablePath() copies the path of the main executable into the
     buffer buf.  The bufsize parameter should initially be the size of the
     buffer.  This function returns 0 if the path was successfully copied.  It
     returns -1 if the buffer is not large enough, and * bufsize is set to the
     size required.  Note that _NSGetExecutablePath() will return "a path" to
     the executable not a "real path" to the executable.  That is, the path
     may be a symbolic link and not the real file. With deep directories the
     total bufsize needed could be more than MAXPATHLEN.
```
meaning, if we invoke our program like `./path/to/../../path/to/program`, it returns a non-real path: `/tmp/./path/to/../../path/to/program`. Calling `realpath()` on the string modified by `_NSGetExecutablePath` normalizes the path. One benefit of using `_NSGetExecutablePath` is that darwin kernel puts it in the process stack right after `envp` array when it is created.

> maybe this API should be implemented as part of the runtime and the runtime should cache the path during startup.

corehost requires this information before loading the runtime. IMO, to unify all four implementations, we can crate an object file out of the slimmest variant and then link in the projects that require it: https://github.com/am11/runtime/commit/4ea037b7e512c4e8687917ad86eec9d9d80f96e2. The caller can cache it after the first call to `GetEntrypointExecutableAbsolutePath`.

Reply:
I think we should discuss the best way to share the PALs separately from this PR. The code sharing is not a new problem. It can be worked on independently and it will need feedback from number of different people. If you would like to work on it, it may be best to start by opening an issue with proposed design and directory structure, etc.

> /tmp/./path/to/../../path/to/program

macOS does concatenation during startup: https://github.com/opensource-apple/dyld/blob/3f928f32597888c5eac6003b9199d972d49857b5/src/dyld.cpp#L4795 . That's good because of it makes the result immutable.

I believe that having the results of this API change when somebody changes current directory is very undesirable characteristics, even if it only happens some of the time. For OSes with this problem, I think this should needs to live in the runtime and the runtime should cache the value during startup. Should all OSes be on this plan or just the ones with this problem?

Another interesting case is what this API should return when the path gets deleted. I guess it should return the original path even if it is no longer valid. It suggests that we should not be calling `realpath` on the result.

Thoughts?

Reply:
> it may be best to start by opening an issue with proposed design and directory structure, etc.

Sure, I will make a proposal. I do not know overarching details about the prospective mini-PAL. I will just base it on this initial use-case.

> That's good because of it makes the result immutable.

was experimenting with this change:

https://github.com/dotnet/runtime/blob/731e011d5964cf2c758f2a6c7342c2ad52860dd4/eng/native/exepath.c#L36-L40

it normalizes `/tmp/./path/to/../../path/to/myprogram` to `/tmp/path/to/myprogram`.

> I believe that having the results of this API change when somebody changes current directory is very undesirable characteristics, even if it only happens some of the time.

Agreed, the value of this property should not change during the execution (since the new value will be pointing to the same file on the filesystem).

Can we change the native method such that it caches the normalized form (agnostic of OS) the first time it is invoked since the process startup and next time returns the cached value without making any system call?

> Another interesting case is what this API should return when the path gets deleted. I guess it should return the original path even if it is no longer valid. It suggests that we should not be calling `realpath` on the result.

Agreed; caching the exec path on first usage and then using the same for lifetime of application is reasonable and something other runtimes are also seem to be doing: https://play.golang.org/p/f9ZmIY_zsSi.